DEF_AUDVIEW = 'DEF_AUDVIEW'
DEF_CAMA = 'DEF_CAMA'
DEF_CAMB = 'DEF_CAMB'
DEF_GRASS = 'DEF_GRASS'
DEF_STADIUM = 'DEF_STADIUM'
DEF_WALL = 'DEF_WALL'
DEF_VISWALL = 'DEF_VISWALL'
NAME_CAMA = 'cam_a'
NAME_CAMB = 'cam_b'
NAME_RECV = 'recv'

# these DEF names are for dynamically created node such as ball and robots
DEF_BALL = 'DEF_BALL'
DEF_BALLSHAPE = 'DEF_BALLSHAPE'
DEF_ORANGESHAPE = 'DEF_ORANGESHAPE'
DEF_ROBOT_PREFIX = 'DEF_ROBOT'

# cams
RECV_PERIOD_MS = 50

SCALE = 1.3
#   Field Dimensions
FIELD_LENGTH = 7.8*SCALE
FIELD_WIDTH = 4.65*SCALE
GOAL_DEPTH = 0.45*SCALE
GOAL_WIDTH = 1.0*SCALE
PENALTY_AREA_DEPTH = 0.9*SCALE
PENALTY_AREA_WIDTH = 1.8*SCALE
GOAL_AREA_DEPTH = 0.4*SCALE
GOAL_AREA_WIDTH = 1.3*SCALE
WALL_THICKNESS = 0.025
CORNER_LENGTH = 0.1*SCALE

#   Ball Dimension
BALL_RADIUS = 0.04
BALL_MASS = 0.0184

NUMBER_OF_ROBOTS = 1
ROBOT_SIZE = [0.15]
ROBOT_HEIGHT = [0.42]
AXLE_LENGTH = [0.14]
ROBOT_BODY_MASS = [2.5]
WHEEL_RADIUS = [0.04]
WHEEL_MASS = [0.15]
MAX_LINEAR_VELOCITY = [1.8]
MAX_TORQUE = [0.8]
PI = 3.14159265358979323846
ROBOT_FORMATION = [
  #  x, y, th - Default Formation
  # [[-1.65*SCALE, -0.19*SCALE, 0]],
  [[-3.7*SCALE, -2.2*SCALE, 0]],
  #  x, y, th - Kickoff Formation
  [[0.4*SCALE,   0.0*SCALE,   PI]]
]
FORMATION_DEFAULT = 0
FORMATION_KICKOFF = 1

# ROBOT_FOUL_POSTURE = [
#  [-4.05*SCALE, -0.85*SCALE, 0],
#  [-4.05*SCALE, -1.15*SCALE, 0],
#  [-4.05*SCALE, -1.45*SCALE, 0],
#  [-4.05*SCALE, -1.75*SCALE, 0],
#  [-4.05*SCALE, -2.05*SCALE, 0],
# ]
BALL_POSTURE = [
 [0.0*SCALE,  0.0*SCALE],
 [-3.25*SCALE,  0.0*SCALE]
]
BALL_DEFAULT = 0
BALL_GOALKICK = 1

# Server settings
SERVER_IP = '127.0.0.1'
SERVER_PORT = 5003
MAX_INSTANCES = 100

KEY_LENGTH = 10

# Game settings
WAIT_READY_MS = 30 * 1000  # ms
# WAIT_KILL_MS = 30 * 1000  # ms
WAIT_STABLE_MS = 1 * 1000  # ms
# WAIT_GOAL_MS = 3 * 1000  # ms
# WAIT_VIEW_MS = 1 * 1000  # ms
WAIT_END_MS = 3 * 1000  # ms
DEFAULT_GAME_TIME_MS = 10 * 1000  # ms
PERIOD_MS = 50  # ms
# PA_THRESHOLD_A = 2  # penalty area robot limit for attacking team
# PA_THRESHOLD_D = 3  # penalty area robot limit for defending team
# DEADLOCK_SENTOUT_NUMBER = 2  # number of robots sent out when a deadlock happens
SENTOUT_DURATION_MS = 5 * 1000  # ms
FALL_TIME_MS = 3 * 1000  # ms
# DEADLOCK_DURATION_MS = 4 * 1000  # ms
# DEADLOCK_THRESHOLD = 0.4  # m/s
# CROWDLOCK_DURATION_MS = 7 * 1000 # m/s
# CROWDLOCK_THRESHOLD = 3
# KICKOFF_TIME_LIMIT_MS = 3 * 1000  # ms
# KICKOFF_BORDER = 0.5*SCALE  # m
# GOALKICK_TIME_LIMIT_MS = 3 * 1000  # ms
# CORNERKICK_TIME_LIMIT_MS = 3 * 1000  # ms
# PENALTYKICK_TIME_LIMIT_MS = 3 * 1000  # ms
# IOPA_TIME_LIMIT_MS = 1 * 1000  # ms
# GK_NIPA_TIME_LIMIT_MS = 1 * 1000  # ms
# NUM_COMMENTS = 3
# MSG_MAX_SIZE = 90000  # bytes
# NONE = 0
# GAME_START = 1
# SCORE_RED_TEAM = 2
# SCORE_BLUE_TEAM = 3
# GAME_END = 4
# DEADLOCK = 5
# GOALKICK = 6
# CORNERKICK = 7
# PENALTYKICK = 8
# HALFTIME = 9
# EPISODE_END = 10
CODEWORDS = [0b000000000, 0b000011111, 0b011100011, 0b101101100, 0b110110101]

# Roles
TEAM_RED = 0
TEAM_BLUE = 1
ROLES = [TEAM_RED, TEAM_BLUE]
TEAMS = [TEAM_RED, TEAM_BLUE]